### ISSUE TITLE
Regression of artifact URL due to change since pip 23.3

### ISSUE THREAD

**Issue by tachikoma-li** [pex-tool/pex#2414]
Looks like there is a "breaking" change from pip in this [commit](https://github.com/pypa/pip/commit/b47f77d330bb7a8642af5490c0f03642f039974c), which applies the build_tracker [conditionally](https://github.com/pypa/pip/commit/b47f77d330bb7a8642af5490c0f03642f039974c#diff-5225c8e359020adb25dfc8c7a505950fd649c6c5775789c6f6517f7913f94542). The result is that there will be no log info from the build tracker module, including the item like [this](https://github.com/pypa/pip/blob/911bf995927aaf56993887225917ba7200df4e7c/src/pip/_internal/operations/build/build_tracker.py#L112) that is used in [pex](https://github.com/pex-tool/pex/blob/36dd2374569b5f3fbef54918e6ba3bd1ff852b61/pex/resolve/locker.py#L314). It seems as a result of this change, the URL for artifacts from private python package index hit the same issue as raised in https://github.com/pex-tool/pex/issues/2156.

It affects pip >= 23.3. Lower versions seem fine with pex generating proper URLs for those artifacts.

As an example, this is the log from pip `23.2.1` for installing `requests`:

<img width="559" alt="image" src="https://github.com/pex-tool/pex/assets/60371004/ec4ca256-8f03-4509-8830-0456844f6640">

and this is from pip `23.3`:

<img width="548" alt="image" src="https://github.com/pex-tool/pex/assets/60371004/161735ae-3780-4674-bd78-1adf6a19dc3e">

**Comment by jsirois** [comment]
@tachikoma-li most useful would be log text (instead of screen shots) and an example of the lock file generated with both versions of Pip. Even better, just the full pex command line (and input requirements files, if any). I'm not exactly sure what problem you are encountering. The #2156 issue was moot, since the querystring that user needed was _dynamic_ and changed over time (an auth token with an expiry). Pex currently absolutely does not support dynamic URLs. Whatever artifact URL is written down in the lockfile when it is created must be good forever after or the lockfile must be regenerated. If your lockfile artifact URLs are missing _static_ querystring data they need, that's different and we can use this issue to track a fix. I look forward to hearing more details about your case.

**Comment by tachikoma-li** [comment]
Hi @jsirois, we have some internal Python packages (in both wheel and sdist format) host in the private PyPI. When we try to generate the lockfile with pex, we have some issues with the URLs for those internal packages. In particular, Pex uses the impermanent URLs for those packages which causes the installation to fail later, and the issue only happens with pip >= 23.3.1. With pip 23.2 (or any lower version), the lockfile can pick up the URL pointing to the private PyPI (instead of the impermanent URLs that have the token for authentication), which allows us to use .netrc to access the PyPI for downloading. For example, this would generate the expected lock
pex3 lock create --use-pip-config <internal_pkg> --pip-version 23.2
but the following fails
pex3 lock create --use-pip-config <internal_pkg> --pip-version 23.3.1
From what I understand, pants needs static URLs for downloading packages in parallel when resolving dependencies. But the behavior seems to be inconsistent with different configurations:
pants can generate expected lock for pip <23.3.1, but fail with higher version of pip
however, with higher versions of pip, pants can also generate the expected lock for sdist, but fail for wheel.
For example,  if we run
pex3 lock create --use-pip-config <internal_pkg> --pip-version 24.0 --only-build <internal_pkg>
we can also get the expected lock.

It seems Pex relies on log from pip to get the information of the artifacts. One thing we notice is this [change](https://github.com/pypa/pip/commit/b47f77d330bb7a8642af5490c0f03642f039974c) from pip can affect the log. For example, all logs related to build tracker are only available for sdist (which is part of the release of pip 23.3), but not for wheel, which seems to explain the inconsistence mentioned above.

Just wondering, is this the intended behavior, it there any plan to fix the issue for wheel (or the make the behavior for sdist similar to wheel?

**Comment by jsirois** [comment]
@tachikoma-li the Pip change you identify may or may not have to do with the issue you see. That said, it should only affect sdists (and VCS URLs and local project paths) and not wheels - exactly the opposite of the problem you're having if I understand correctly.

> Just wondering, is this the intended behavior, is there any plan to fix the issue for wheel (or the make the behavior for sdist similar to wheel?

The Pex integration tests are extensive and include locks of wheels, sdists, VCS URLs and even local projects across the full range of supported Pip versions. As such, you're hitting a corner here that almost certainly involves the specifics of the private PyPI setup you use. Note that Pex CI itself uses a "private" devpi mirror instead of PyPI directly - although there is no auth used.

More to the point, I really need to be able to reproduce your issue to solve it. Can you please provide all of the details from either the "Indirect" or "Direct" lists below?

Indirect:
+ The full Pip log (add `--preserve-pip-download-log` to your `pex3 lock create` command line, observe a line like: `pex: Preserving `pip download` log at /tmp/pex-pip-log.1zil7ou6/pip.log` and then grab that fil and search / replace any sensitive info you need to redact.
+ The full before (good) and after (bad) lock files - again redacted as needed.

Direct:
+ The private PyPI server implementation you use and its setup.
+ How you think you're passing credentials to the private PyPI server (you mention `~/.netrc` but you also pass `--use-pip-config` to Pex which is potentially another source of auth info).

**Comment by tachikoma-li** [comment]
@jsirois Unfortunately, we can't provide the full pip log for now. Our internal server will redirect to AWS to fetch the packages when recieving download requests, which seems to have the same behavior as `devpi`. We tried `devpi` today and did find something similar to our case, which hopefully can help clarify the issue. 
Steps to reproduce:
- define the index in ~/.pip/pip.conf as
```shell
[global]
index-url=https://m.devpi.net/root/pypi/+simple/
```
 - run `pex3 lock create --use-pip-config wheel --pip-version 23.3.1` which outputs
 ```shell
{"allow_builds": true, "allow_prereleases": false, "allow_wheels": true, "build_isolation": true, "constraints": [], "locked_resolves": [{"locked_requirements": [{"artifacts": [{"algorithm": "sha256", "hash": "55c570405f142630c6b9f72fe09d9b67cf1477fcf543ae5b8dcb1f5b7377da81", "url": "https://files.pythonhosted.org/packages/7d/cd/d7460c9a869b16c3dd4e1e403cce337df165368c71d6af229a74699622ce/wheel-0.43.0-py3-none-any.whl"}], "project_name": "wheel", "requires_dists": ["pytest>=6.0.0; extra == \"test\"", "setuptools>=65; extra == \"test\""], "requires_python": ">=3.8", "version": "0.43.0"}], "platform_tag": ["cp310", "cp310", "macosx_14_0_arm64"]}], "only_builds": [], "only_wheels": [], "path_mappings": {}, "pex_version": "2.3.1", "pip_version": "23.3.1", "prefer_older_binary": false, "requirements": ["wheel"], "requires_python": [], "resolver_version": "pip-2020-resolver", "style": "strict", "target_systems": [], "transitive": true, "use_pep517": null}
```
Note that the URL is the redirected one.

-  run `pex3 lock create --use-pip-config wheel --only-build wheel --pip-version 23.3.1` for `sdist`, which outputs:
```shell
{"allow_builds": true, "allow_prereleases": false, "allow_wheels": true, "build_isolation": true, "constraints": [], "locked_resolves": [{"locked_requirements": [{"artifacts": [{"algorithm": "sha256", "hash": "465ef92c69fa5c5da2d1cf8ac40559a8c940886afcef87dcf14b9470862f1d85", "url": "https://m.devpi.net/root/pypi/%2Bf/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz"}], "project_name": "wheel", "requires_dists": ["pytest>=6.0.0; extra == \"test\"", "setuptools>=65; extra == \"test\""], "requires_python": ">=3.8", "version": "0.43.0"}], "platform_tag": ["cp310", "cp310", "macosx_14_0_arm64"]}], "only_builds": ["wheel"], "only_wheels": [], "path_mappings": {}, "pex_version": "2.3.1", "pip_version": "23.3.1", "prefer_older_binary": false, "requirements": ["wheel"], "requires_python": [], "resolver_version": "pip-2020-resolver", "style": "strict", "target_systems": [], "transitive": true, "use_pep517": null}
```
Now note the URL is not the redirected one.

- run with pip 23.2 `pex3 lock create --use-pip-config wheel --pip-version 23.2`  which outputs
```shell
{"allow_builds": true, "allow_prereleases": false, "allow_wheels": true, "build_isolation": true, "constraints": [], "locked_resolves": [{"locked_requirements": [{"artifacts": [{"algorithm": "sha256", "hash": "55c570405f142630c6b9f72fe09d9b67cf1477fcf543ae5b8dcb1f5b7377da81", "url": "https://m.devpi.net/root/pypi/%2Bf/55c/570405f142630/wheel-0.43.0-py3-none-any.whl"}], "project_name": "wheel", "requires_dists": ["pytest>=6.0.0; extra == \"test\"", "setuptools>=65; extra == \"test\""], "requires_python": ">=3.8", "version": "0.43.0"}], "platform_tag": ["cp310", "cp310", "macosx_14_0_arm64"]}], "only_builds": [], "only_wheels": [], "path_mappings": {}, "pex_version": "2.3.1", "pip_version": "23.2", "prefer_older_binary": false, "requirements": ["wheel"], "requires_python": [], "resolver_version": "pip-2020-resolver", "style": "strict", "target_systems": [], "transitive": true, "use_pep517": null}`
```
Note that the URL is not the redirected one.

This is very close to our issue where the redirected URL is impermanent with auth tokens and causing trouble, while the original URL would work. And this would only affect pip version >= 23.3.1 (excluding the sdist case).

**Comment by jsirois** [comment]
@tachikoma-liI thanks, I think your last post starts to help reveal the issue here - its use of `~/.pip/pip.conf` / `--use-pip-config`. The `--use-pip-config` flag prevents Pex from running Pip with `--isolated`, so Pip can read its `~/.pip/pip.conf` config, but, importantly, Pex does not read `~/.pip/pip.conf`. So when you tell Pip (and not Pex), that the index URL is https://m.devpi.net/root/pypi/+simple/, Pex still uses the default index URL of PyPI. As things stand, you need to duplicate that piece of config information so Pex knows. Like so:
```
pex3 lock create --use-pip-config --no-pypi --index https://m.devpi.net/root/pypi/+simple/ ...
```

This is a bug, but a long standing one (you can also specify extra indexes in a requirements.txt file and Pip will use those but Pex will not known about them - same bug). The idea of introducing --use-pip-config (which was introduce in September 2023 in v2.1.147) was to enable things like [Pip keyring support](https://pip.pypa.io/en/stable/topics/authentication/#keyring-support).

@tachikoma-li now that you know you must tell Pex directly about the indexes to use, does doing so fix your problem?

**Comment by jsirois** [comment]
To explain what I think the issue really is here:

You're using Pants to run pex and Pants does this:
+ https://github.com/pantsbuild/pants/blob/aaf94df44063168771ade199da52a8fc6c73a84f/src/python/pants/backend/python/util_rules/pex_requirements.py#L354-L360
+ https://github.com/pantsbuild/pants/blob/aaf94df44063168771ade199da52a8fc6c73a84f/src/python/pants/backend/python/subsystems/repos.py#L15-L25

As such Pants ends up launching Pex like so:
```
pex3 lock create --no-pypi --index https://pypi.org/simple/ --use-pip-config ...
```

Which mean Pex ends up launching Pip like so:
```
pip download --index-url https://pypi.org/simple/ ...
```

So now Pip is configured with 2 `--index-url`s - the one from the command line and the other from `~/.pip/pip.conf`. When faced with multiple indexes, Pip is notorious for not defining a rule for the precedence they will be used in and that precedence has, in fact, shifted over time (see https://github.com/pypa/pip/issues/9484#issuecomment-764470437 for example). This shifting of precedence over Pip releases (and not the logging patch you suspected) likely explains why things worked for you under one Pip version and not another.

**Comment by tachikoma-li** [comment]
Hi @jsirois, thanks for the quick reply!
I just did some tests locally with a single index, but seemed still to hit the same issue:
- run `pex3 lock create --index https://m.devpi.net/root/pypi/+simple/ wheel --pip-version 23.3.1` which outputs:
```
{"allow_builds": true, "allow_prereleases": false, "allow_wheels": true, "build_isolation": true, "constraints": [], "locked_resolves": [{"locked_requirements": [{"artifacts": [{"algorithm": "sha256", "hash": "55c570405f142630c6b9f72fe09d9b67cf1477fcf543ae5b8dcb1f5b7377da81", "url": "https://files.pythonhosted.org/packages/7d/cd/d7460c9a869b16c3dd4e1e403cce337df165368c71d6af229a74699622ce/wheel-0.43.0-py3-none-any.whl"}], "project_name": "wheel", "requires_dists": ["pytest>=6.0.0; extra == \"test\"", "setuptools>=65; extra == \"test\""], "requires_python": ">=3.8", "version": "0.43.0"}], "platform_tag": ["cp310", "cp310", "macosx_14_0_arm64"]}], "only_builds": [], "only_wheels": [], "path_mappings": {}, "pex_version": "2.3.1", "pip_version": "23.3.1", "prefer_older_binary": false, "requirements": ["wheel"], "requires_python": [], "resolver_version": "pip-2020-resolver", "style": "strict", "target_systems": [], "transitive": true, "use_pep517": null}
```
with a redirected URL.

- run `pex3 lock create --index https://m.devpi.net/root/pypi/+simple/ wheel --pip-version 23.3.1 --only-build wheel` which outputs 
 ```
{"allow_builds": true, "allow_prereleases": false, "allow_wheels": true, "build_isolation": true, "constraints": [], "locked_resolves": [{"locked_requirements": [{"artifacts": [{"algorithm": "sha256", "hash": "465ef92c69fa5c5da2d1cf8ac40559a8c940886afcef87dcf14b9470862f1d85", "url": "https://m.devpi.net/root/pypi/%2Bf/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz"}], "project_name": "wheel", "requires_dists": ["pytest>=6.0.0; extra == \"test\"", "setuptools>=65; extra == \"test\""], "requires_python": ">=3.8", "version": "0.43.0"}], "platform_tag": ["cp310", "cp310", "macosx_14_0_arm64"]}], "only_builds": ["wheel"], "only_wheels": [], "path_mappings": {}, "pex_version": "2.3.1", "pip_version": "23.3.1", "prefer_older_binary": false, "requirements": ["wheel"], "requires_python": [], "resolver_version": "pip-2020-resolver", "style": "strict", "target_systems": [], "transitive": true, "use_pep517": null}
```
with the original URL.

- run `pex3 lock create --index https://m.devpi.net/root/pypi/+simple/ wheel --pip-version 23.2` which outputs 
 ```
{"allow_builds": true, "allow_prereleases": false, "allow_wheels": true, "build_isolation": true, "constraints": [], "locked_resolves": [{"locked_requirements": [{"artifacts": [{"algorithm": "sha256", "hash": "55c570405f142630c6b9f72fe09d9b67cf1477fcf543ae5b8dcb1f5b7377da81", "url": "https://m.devpi.net/root/pypi/%2Bf/55c/570405f142630/wheel-0.43.0-py3-none-any.whl"}], "project_name": "wheel", "requires_dists": ["pytest>=6.0.0; extra == \"test\"", "setuptools>=65; extra == \"test\""], "requires_python": ">=3.8", "version": "0.43.0"}], "platform_tag": ["cp310", "cp310", "macosx_14_0_arm64"]}], "only_builds": [], "only_wheels": [], "path_mappings": {}, "pex_version": "2.3.1", "pip_version": "23.2", "prefer_older_binary": false, "requirements": ["wheel"], "requires_python": [], "resolver_version": "pip-2020-resolver", "style": "strict", "target_systems": [], "transitive": true, "use_pep517": null}
```
with the original URL. That is, it seems the issue remains for a single index.

**Comment by jsirois** [comment]
@tachikoma-li `--index` is _not_ single index, it adds an index to the default (PyPI). Use `--no-pypi --index https://m.devpi.net/root/pypi/+simple/`.

Perhaps this is not clear from the `--help`:
```
pex --help
...
 --pypi, --no-pypi, --no-index
                        Whether to use PyPI to resolve dependencies. (default:
                        True)
  -f PATH/URL, --find-links PATH/URL, --repo PATH/URL
                        Additional repository path (directory or URL) to look
                        for requirements. (default: None)
  -i URL, --index URL, --index-url URL
                        Additional cheeseshop indices to use to satisfy
                        requirements. (default: None)
...
```

**Comment by tachikoma-li** [comment]
@jsirois, thanks. Just tried with `--no-pypi --index`, but still got the same result as reported above.
That is, `pex3 lock create --no-pypi --index https://m.devpi.net/root/pypi/+simple/ wheel --pip-version 23.3.1 --only-build wheel` outputs the right URL, but `pex3 lock create --no-pypi --index https://m.devpi.net/root/pypi/+simple/ wheel --pip-version 23.3.1` outputs the redirected one.

**Comment by jsirois** [comment]
@tachikoma-li Ok, great - thanks. I repro that at the 23.2 / 23.3.1 boundary. This is enough to work with now.

**Comment by jsirois** [comment]
The fix is released in 2.3.3: https://github.com/pex-tool/pex/releases/tag/v2.3.3
